### DISCUSSION TITLE
代码动态切换主题时，动画storyboard的颜色不能动态调整

### DISCUSSION THREAD

**Discussion by lisery454** [BYJRK/DotNet-Discussions#247] [WPF 开发]
最近做项目想做一个动态修改颜色主题的功能，实现方式是用代码替换`App.xaml`里的资源字典文件，然后组件使用`DynamicResources`绑定`Brush`和`Color`。
这样做一般是没什么问题。但是我又想做一些动画，比如按钮鼠标移上去颜色会渐变。
但是我试了之后发现不行，会报`InvalidOperationException: 无法冻结此 Storyboard 时间线树供跨线程使用`的异常，我查了之后知道了动画的`To`和`From`不能使用`Binding`和`DynamicResource`，只能绑定静态常量。
然后我查了一些资料，有种方法弄一个0到1变量进行`Converter`映射颜色，但是我还是做不到动态改变这个动画颜色。
我是在`ControlTemplate`的`triggers`里写的动画。
不知道这个问题要怎么解决。

**Comment by BYJRK** [comment]
动画的 From、To 之类因为要求冻结，所以是无法通过绑定来实现的。你所说的需求通常只能用代码后台动态生成动画来实现。From、To 的值可以用 DynamicResource 来给出，然后在代码后台通过 FindResource 的方式拿到，并声明动画

**Reply by lisery454** [reply]
我是在一个资源字典文件里声明这个Style，然后在style的template里面写动画，这样的话我没有后台代码来进行绑定啊。还是说弄一个全局的静态变量，通过x:Static来访问？刚学不是很懂。

**Comment by SlimeNull** [comment]
写俩storyboard，然后用trigger做切换？
